fixit/rules/chained_instance_check.py

```diff
@@ -0,0 +1,114 @@
+    can take a tuple of types and check whether given target suits
+    any of them. Rather than chaining multiple ``isinstance`` calls
+    with a boolean-or operation, a single ``isinstance`` call where
+    the second argument is a tuple of all types can be used.
+    """
+
+    MESSAGE: str = (
+        "Multiple isinstance calls with the same target but "
+        + "different types can be collapsed into a single call "
+        + "with a tuple of types."
+    )
+
+    VALID = [
+        Valid("foo() or foo()"),
+        Valid("foo(x, y) or foo(x, z)"),
+        Valid("foo(x, y) or foo(x, z) or foo(x, q)"),
+        Valid("isinstance() or isinstance()"),
+        Valid("isinstance(x) or isinstance(x)"),
+        Valid("isinstance(x, y) or isinstance(x)"),
+        Valid("isinstance(x) or isinstance(x, y)"),
+        Valid("isinstance(x, y) or isinstance(t, y)"),
+        Valid("isinstance(x, y) and isinstance(x, z)"),
+        Valid("isinstance(x, y) or isinstance(x, (z, q))"),
+        Valid("isinstance(x, (y, z)) or isinstance(x, q)"),
+    ]
+    INVALID = [
+        Invalid(
+            "isinstance(x, y) or isinstance(x, z)",
+            expected_replacement="isinstance(x, (y, z))",
+        ),
+        Invalid(
+            "isinstance(x, y) or isinstance(x, z) or isinstance(x, q)",
+            expected_replacement="isinstance(x, (y, z, q))",
+        ),
+        Invalid(
+            "isinstance(x, y) or isinstance(x, z) or isinstance(x, q) or isinstance(x, w)",
+            expected_replacement="isinstance(x, (y, z, q, w))",
+        ),
+    ]
+
+    def __init__(self, context: CstContext) -> None:
+        super().__init__(context)
+
+        # Since we already unwrap a boolean op's
+        # children
+        self.seen_boolean_operations: Set[cst.BooleanOperation] = set()
+
+    def visit_BooleanOperation(self, node: cst.BooleanOperation) -> None:
+        # Initially match with a partial pattern (in order to ensure
+        # we have enough args to construct more accurate / advanced
+        # pattern).
+        if node not in self.seen_boolean_operations and m.matches(
+            node,
+            m.BooleanOperation(right=m.Call(args=[m.AtLeastN(n=1)]), operator=m.Or()),
+        ):
+            args = self._collect_isinstance_args(node)
+            if args is not None:
+                right = cst.ensure_type(node.right, cst.Call)
+                elements = [cst.Element(arg.value) for arg in args]
+                new_isinstance_call = right.with_deep_changes(
+                    old_node=right.args[1], value=cst.Tuple(elements)
+                )
+                self.report(node, replacement=new_isinstance_call)
+
+    def _collect_isinstance_args(
+        self, node: cst.BooleanOperation
+    ) -> Optional[List[cst.Arg]]:
+        target = cst.ensure_type(node.right, cst.Call).args[0].value
+        expected_call = m.Call(
+            func=m.Name(value="isinstance"),
```

Comment:
To make this rule more robust, we'll want to make sure the `isinstance` is `builtins.isinstance` rather than a local defined function or an import alias.
We can use `QualifiedNameProvider.has_name` to check this.
Add a metadata dependency `METADATA_DEPENDENCIES = (QualifiedNameProvider,)` to the lint rule and then 
use the has_name helper to check it, e.g.
https://github.com/Instagram/Fixit/blob/18365db024258817f7f30006886c14ce8f07bbd5/fixit/rules/cls_in_classmethod.py#L225-L230

Reference for metadata API.
https://libcst.readthedocs.io/en/latest/metadata_tutorial.html
https://libcst.readthedocs.io/en/latest/metadata.html

Reply:
Yeah, makes sense!